flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -1243,7 +1243,7 @@ private void completePendingCheckpoint(PendingCheckpoint pendingCheckpoint)
             // the pending checkpoint must be discarded after the finalization
             Preconditions.checkState(pendingCheckpoint.isDisposed() && completedCheckpoint != null);
 
-            if (!props.isSavepoint()) {
+            if (!props.isSavepoint() || props.isSynchronous()) {
                 lastSubsumed =
                         addCompletedCheckpointToStoreAndSubsumeOldest(
```

Comment:
Should we subsume anything in this case? Also isn't this braking savepoint ownership?

Reply:
1. For the subsumption, we do have a logic that will keep at least one other checkpoint (not-savepoint), so I think that's fine. (see: `CheckpointSubsumeHelper#subsume`).
2. As for the ownership, there are two aspects here. First, we can say savepoint (stop-with-savepoint) is under Flink's control until the job has finished. Secondly, this reverts to the behaviour of <=1.14, the better handling we discussed should be implemented in https://issues.apache.org/jira/browse/FLINK-26683